src/object/srv_ec_aggregate.c

```diff
@@ -0,0 +1,1012 @@
+			d_list_del(&agg_extent->ae_link);
+			D_FREE_PTR(agg_extent);
+		}
+	}
+	agg_entry->ae_cur_stripe.as_offset = 0U;
+	/* Account for carry over. */
+	if (ptail)
+		agg_entry->ae_cur_stripe.as_stripenum++;
+	else
+		agg_entry->ae_cur_stripe.as_hi_epoch = 0UL;
+	agg_entry->ae_cur_stripe.as_stripe_fill = ptail;
+}
+
+/* Returns the stripe number for the stripe containing ex_lo.
+ */
+static inline daos_off_t
+agg_stripenum(struct ec_agg_entry *entry, daos_off_t ex_lo)
+{
+	return ex_lo / obj_ec_stripe_rec_nr(entry->ae_oca);
+}
+
+/* Call back for the nested iterator used to find the parity for a stripe.
+ */
+static int
+agg_recx_iter_pre_cb(daos_handle_t ih, vos_iter_entry_t *entry,
+		     vos_iter_type_t type, vos_iter_param_t *param,
+		     void *cb_arg, unsigned int *acts)
+{
+	struct ec_agg_entry	*agg_entry = (struct ec_agg_entry *)cb_arg;
+	int			 rc = 0;
+
+	D_ASSERT(type == VOS_ITER_RECX);
+	D_ASSERT(entry->ie_recx.rx_idx == (PARITY_INDICATOR |
+		(agg_entry->ae_cur_stripe.as_stripenum *
+			agg_entry->ae_oca->u.ec.e_len)));
+	agg_entry->ae_par_extent.ape_recx = entry->ie_recx;
+	agg_entry->ae_par_extent.ape_epoch = entry->ie_epoch;
+	return rc;
+}
+
+enum agg_iov_entry {
+	AGG_IOV_DATA	= 0,
+	AGG_IOV_ODATA,
+	AGG_IOV_PARITY,
+	AGG_IOV_DIFF,
+	AGG_IOV_CNT,
+};
+
+/* Allocates an sgl iov_buf at iov_entry offset in the array.
+ */
+static int
+agg_alloc_buf(d_sg_list_t *sgl, size_t ent_buf_len, unsigned int iov_entry,
+	      bool align_data)
+{
+	int		 rc = 0;
+
+	if (align_data) {
+		D_FREE(sgl->sg_iovs[iov_entry].iov_buf);
+		sgl->sg_iovs[iov_entry].iov_buf =
+			aligned_alloc(32, ent_buf_len);
+		if (sgl->sg_iovs[iov_entry].iov_buf == NULL) {
+			rc = -DER_NOMEM;
+			goto out;
+		}
+	} else {
+		unsigned int *buf = NULL;
+
+		D_REALLOC(buf, sgl->sg_iovs[iov_entry].iov_buf, ent_buf_len);
+		 if (buf == NULL) {
+			rc = -DER_NOMEM;
+			goto out;
+		 }
+		sgl->sg_iovs[iov_entry].iov_buf = buf;
+	}
+	sgl->sg_iovs[iov_entry].iov_len = ent_buf_len;
+	sgl->sg_iovs[iov_entry].iov_buf_len = ent_buf_len;
+out:
+	return rc;
+}
+
+/* Prepares the SGL used for VOS I/O and peer target I/O (update needed
+ * to use library functions (e.g., d_iov_set).
+ *
+ * This function is a no-op if entry's sgl is sufficient for the current
+ * object class.
+ *
+ */
+static int
+agg_prep_sgl(struct ec_agg_entry *entry)
+{
+	size_t		 data_buf_len, par_buf_len;
+	unsigned int	 len = entry->ae_oca->u.ec.e_len;
+	unsigned int	 k = entry->ae_oca->u.ec.e_k;
+	unsigned int	 p = entry->ae_oca->u.ec.e_p;
+	int		 rc = 0;
+
+	if (entry->ae_sgl.sg_nr == 0) {
+		D_ALLOC_ARRAY(entry->ae_sgl.sg_iovs, AGG_IOV_CNT);
+		if (entry->ae_sgl.sg_iovs == NULL) {
+			rc = -DER_NOMEM;
+			goto out;
+		}
+		entry->ae_sgl.sg_nr = AGG_IOV_CNT;
+	}
+	D_ASSERT(entry->ae_sgl.sg_nr == AGG_IOV_CNT);
+	data_buf_len = len * k * entry->ae_rsize;
+	if (entry->ae_sgl.sg_iovs[AGG_IOV_DATA].iov_buf_len < data_buf_len) {
+		rc = agg_alloc_buf(&entry->ae_sgl, data_buf_len, AGG_IOV_DATA,
+				   true);
```

Comment:
hmm, I thought cell size must be 32 bytes aligned?

Reply:
The data buffers (AGG_IOV_DATA, AGG_IOV_ODATA, and AGG_IOV_DIFF) have to be 32-byte aligned to enable incremental update of parity.  The parity buffer (AGG_IOV_PARITY) does not need the alignment.  the function agg_alloc_buf aligns if the last arg is true.  realloc can be more efficient, so the  agg_alloc_buf funthe ctions uses realloc for the parity buffer.